### PR TITLE
[Done] API for dist train

### PR DESCRIPTION
**Pull request by typhoonzero**
Fix https://github.com/PaddlePaddle/Paddle/issues/6286

Multi trainers still have some problem, will refine in next PR.

### CODE REVIEW

**Comment by Yancey1989** [comment]
Maybe the trainer can not connect to the pserver, too much TIME-WAIT TCP state, and no any logs
```bash
λ 3fab79b2d17f / netstat -anp |grep 6174
grep: warning: GREP_OPTIONS is deprecated; please use an alias or script
tcp        0      0 127.0.0.1:6174          0.0.0.0:*               LISTEN      94/python
tcp        0      0 127.0.0.1:39128         127.0.0.1:6174          TIME_WAIT   -
tcp        0      0 127.0.0.1:39654         127.0.0.1:6174          TIME_WAIT   -
tcp        0      0 127.0.0.1:39538         127.0.0.1:6174          TIME_WAIT   -
tcp        0      0 127.0.0.1:39504         127.0.0.1:6174          TIME_WAIT   -
tcp        0      0 127.0.0.1:39548         127.0.0.1:6174          TIME_WAIT   -
tcp        0      0 127.0.0.1:39160         127.0.0.1:6174          TIME_WAIT   -
tcp        0      0 127.0.0.1:39238         127.0.0.1:6174          TIME_WAIT   -
tcp        0      0 127.0.0.1:38952         127.0.0.1:6174          TIME_WAIT   -
tcp        0      0 127.0.0.1:39592         127.0.0.1:6174          TIME_WAIT   -
tcp        0      0 127.0.0.1:39332         127.0.0.1:6174          TIME_WAIT   -
tcp        0      0 127.0.0.1:39464         127.0.0.1:6174          TIME_WAIT   -
tcp        0      0 127.0.0.1:39090         127.0.0.1:6174          TIME_WAIT   -
...
```

- start pserver
```bash
> PSERVER=127.0.0.1:6174 python notest_recognize_digits_conv_dist.py
```
- start trainer
```bash
> python notest_recognize_digits_conv_dist.py
```

**Comment by Yancey1989** [comment]
Create another PR to this branch :  https://github.com/typhoonzero/Paddle/pull/2
1. Fixed unit test failed, the `sum` op should have Input(X).
1. Remaining  problem: 
```bash
I1220 12:13:03.567848 24671 executor.cc:129] Op(sum), inputs:{X[x0, x1]}, outputs:{Out[Out]}.
I1220 12:13:03.567870 24671 executor.cc:129] Op(sum), inputs:{X[x0, x1]}, outputs:{Out[Out]}.
I1220 12:13:03.567893 24671 executor.cc:129] Op(sum), inputs:{X[x0, x1]}, outputs:{Out[Out]}.
I1220 12:13:03.567896 24670 scope.cc:28] Destroy variable Out
I1220 12:13:03.567917 24671 executor.cc:129] Op(sum), inputs:{X[x0, x1]}, outputs:{Out[Out]}.
I1220 12:13:03.567924 24670 scope.cc:28] Destroy variable x1
I1220 12:13:03.567939 24670 scope.cc:28] Destroy variable x0
I1220 12:13:03.567942 24671 executor.cc:129] Op(sum), inputs:{X[x0, x1]}, outputs:{Out[Out]}.
terminate called without an active exception
I1220 12:13:03.567975 24671 executor.cc:129] Op(sum), inputs:{X[x0, x1]}, outputs:{Out[Out]}.
Aborted (core dumped)
```

Maybe trainer/pserver need graceful shutdown.

File: paddle/operators/recv_op.cc [hunk 1]
```diff
@@ -61,29 +61,74 @@ class RecvOp : public framework::OperatorBase {
...
-    // Run sub graph to get optimized tensor
-    executor.Run(program, &recv_scope, 0, /*global_block*/
-                 false /*create_local_scope*/);
-
-    auto *out_var = recv_scope.FindVar("Out");
-    // push back
-    rpc_service_->Push(out_var->Get<framework::LoDTensor>());
+    rpc_service_->SetScope(&recv_scope);
+    auto param_list = Attr<std::vector<std::string>>("ParamList");
+    auto grad_list = Attr<std::vector<std::string>>("GradList");
```
[the remark this answers is not in the document]
Comment (Yancey1989) [reply paddle/operators/recv_op.cc]: I think the `grad_list` and `param_list` are used to ensure pserver received a pair of parameter vs. gradient, and maybe we could do this in an Op which named `SyncWait`(or other name): checking the parameters vs. gradients in the blocking queue until received all tensors.

File: paddle/operators/send_op.cc [hunk 2]
```diff
@@ -34,45 +34,56 @@ class SendOp : public framework::OperatorBase {
...
-    auto iname = Input("X");
-    auto oname = Output("Out");
-    // TODO(typhoonzero): currently it's non-blocking,
-    // should block until server responds.
-    bool ret = client_->SendVariable(scope, iname, oname);
-    if (!ret) {
-      LOG(ERROR) << "send variable error";
+    auto ins = Inputs("X");
+    std::vector<std::string> epmap = Attr<std::vector<std::string>>("epmap");
+    // TODO(typhoonzero): use async calls to send multiple variable asyncly.
```
[the remark this answers is not in the document]
Comment (Yancey1989) [reply paddle/operators/send_op.cc]: Sorry, I'm not sure the meaning of `one send-recv pair for each variable`,  does it need a independent port for each variable?

[the remark this answers is not in the document]
Comment (helinwang) [reply paddle/operators/send_op.cc]: I was thinking there is a singleton service that listens on one port, and all send-recv ops go through that port.

File: paddle/operators/recv_op.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (helinwang) [reply paddle/operators/recv_op.cc]: I think if we use one send-recv op for one var, then we don't need to maintain two lists. Another benefit is here: https://github.com/PaddlePaddle/Paddle/pull/6297#discussion_r158121595

[the remark this answers is not in the document]
Comment (Yancey1989) [reply paddle/operators/recv_op.cc]: Sure, but I think we also need a global Blocking Queue to store the variant so that recv op will fetch variant from the queue.

[the remark this answers is not in the document]
Comment (typhoonzero) [reply paddle/operators/recv_op.cc]: @helinwang Indeed, `param_list` is not used in fact. The reason of we need to know the variable names is that we need to merge variables from multiple workers. The variable names from multiple workers are the same, so we need to rename them by adding numberic suffixes: `xxx@GRAD.trainer_0` to `xxx@GRAD.trainer_n` then the `sum_op` will use them as input.

The rename operation must happen before the sub program runs, and only `recv_op` can know when the variable arrives. I can't find a better way right now. But still, can make the `recv_op` more general by change the names.

[the remark this answers is not in the document]
Comment (typhoonzero) [reply paddle/operators/recv_op.cc]: @Yancey1989 `param_list` is in fact not used for now, will remove.

The idea of using a queue before sending variables is great that we can send the variable parallelly right after the variable is ready to gain some performance.

[the remark this answers is not in the document]
Comment (helinwang) [reply paddle/operators/recv_op.cc]: @typhoonzero I think we can merge this PR soon, this is great work we can merge in to the develop branch to start iterate.
In the same time, in my mind send-recv op is lower level than merging variables, it should be used by the merging variable logic, rather than controlling the merging variable logic.
You mentioned we need to have more "temp" variables, maybe they should be created by the transpiler, and send-recv op does not need to know it. For example, I saw `snprintf(ret, sizeof(ret), "%s.trainer_%d", varname.c_str()` in `recv_op.cc`, maybe it should be in the transpiler.

File: paddle/operators/recv_op.cc [hunk 3]
```diff
@@ -61,29 +61,74 @@ class RecvOp : public framework::OperatorBase {
...
+    auto trainer_count = Attr<int>("Trainers");
+    size_t param_count = param_list.size();
+    rpc_service_->Start();
```
[the remark this answers is not in the document]
Comment (typhoonzero) [reply paddle/operators/recv_op.cc]: `recv_op`'s `Run` will only be called once, there's an inner event loop inside it. This will be changed to use while op in later versions: https://github.com/PaddlePaddle/Paddle/issues/6508

[the remark this answers is not in the document]
Comment (helinwang) [reply paddle/operators/recv_op.cc]: I think while loop is optional for the user to use. It's still very common for the user to used Python for loop and call run multiple times. So probably we need to have recv's run low overhead.
In my mind the recv op should no call while op, it should be blocked and the executor run the next op when it's done.

File: python/paddle/v2/fluid/distribute_transpiler.py [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+            Use different methods to split trainable varialbles to different
+            parameter servers.
+
+            Example to run:
+
+            exe = fluid.Executor(place)
+            t = fluid.DistributeTranspiler()
+            t.transpile(optimize_ops, params_grads, pservers="127.0.0.1:6174", trainers=1)
+
+            pserver_endpoint = os.getenv("PSERVER")
```
[the remark this answers is not in the document]
Comment (typhoonzero) [reply python/paddle/v2/fluid/distribute_transpiler.py]: Sure, added this to project.

File: paddle/operators/recv_op.cc [hunk 5]
```diff
@@ -61,29 +61,74 @@ class RecvOp : public framework::OperatorBase {
...
+        if (trainer_count > 1) {
+          grad_var_name = this->GetGradVarNameForTrainer(grad_var_name);
+        }
+
+        auto *var = recv_scope.Var(grad_var_name);
+        auto *tensor = var->GetMutable<framework::LoDTensor>();
+        // FIXME(typhoonzero): do not copy
+        framework::CopyFrom(v.second, dev_ctx.GetPlace(), dev_ctx, tensor);
+      }
+      rpc_service_->Start();
```
[the remark this answers is not in the document]
Comment (typhoonzero) [reply paddle/operators/recv_op.cc]: It's just a state reset, renamed to `Reset`.

File: paddle/operators/recv_op.cc [hunk 6]
```diff
@@ -61,29 +61,74 @@ class RecvOp : public framework::OperatorBase {
...
+        std::string param_var_name;
+        if (it != grad_list.end()) {
+          param_var_name = param_list[it - grad_list.begin()];
+        }
+        VLOG(3) << "recved grad: " << grad_var_name
+                << " updating param: " << param_var_name;
+        auto *merged_grad = recv_scope.FindVar(grad_var_name);
+        if (merged_grad == nullptr) {
+          // create output of merged var.
+          auto merged_var = recv_scope.Var(grad_var_name);
```
[the remark this answers is not in the document]
Comment (typhoonzero) [reply paddle/operators/recv_op.cc]: Merging OP is currently added by `transpiler`, recv_op need to rename the variables sent by workers. Will refine this in following PRs.

File: paddle/operators/recv_op.cc [hunk 7]
```diff
@@ -61,29 +61,74 @@ class RecvOp : public framework::OperatorBase {
...
+      // the gradient arrives, just add suffix 0~n then average the gradient.
+      for (size_t i = 0; i < param_count * trainer_count; ++i) {
+        // blocking get one var from client.
+        const detail::TensorWithName &v = rpc_service_->Get();
+        auto grad_var_name = v.first;
+        auto it = std::find(grad_list.begin(), grad_list.end(), grad_var_name);
+        std::string param_var_name;
+        if (it != grad_list.end()) {
+          param_var_name = param_list[it - grad_list.begin()];
+        }
```
[the remark this answers is not in the document]
Comment (typhoonzero) [reply paddle/operators/recv_op.cc]: Done.

File: paddle/operators/recv_op.cc [hunk 8]
```diff
@@ -61,29 +61,74 @@ class RecvOp : public framework::OperatorBase {
...
+    auto param_list = Attr<std::vector<std::string>>("ParamList");
+    auto grad_list = Attr<std::vector<std::string>>("GradList");
+    auto trainer_count = Attr<int>("Trainers");
+    size_t param_count = param_list.size();
+    rpc_service_->Start();
+    // TODO(typhoonzero): change this to a while_op for every cluster-batch.
+    while (true) {
+      // Get from multiple trainers, we don't care about order in which
+      // the gradient arrives, just add suffix 0~n then average the gradient.
+      for (size_t i = 0; i < param_count * trainer_count; ++i) {
```
[the remark this answers is not in the document]
Comment (typhoonzero) [reply paddle/operators/recv_op.cc]: Can we call this `fan_in` in general?

File: paddle/operators/recv_op.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (typhoonzero) [reply paddle/operators/recv_op.cc]: Agree!

File: paddle/operators/recv_op.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (helinwang) [reply paddle/operators/recv_op.cc]: Sure, eventually we can move to one recv op receiving one variable.

File: paddle/operators/send_op.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (Yancey1989) [reply paddle/operators/send_op.cc]: Got it, thanks.
